tune-threads/tune-threads.lisp

```diff
@@ -0,0 +1,169 @@
+;;;; tune-threads/tune-threads.lisp
+;;;;
+;;;; Author: John Lapeyre
+
+(in-package #:qvm-tune-threads)
+
+(defstruct pre-timing
+  "Data characterizing coarse timing of quil program."
+  (init-num-trials 1 :read-only t)
+  (num-trials 1 :read-only t)
+  (trial-time 1.0 :read-only t))
+
+(defparameter *minimum-target-time* 0.1
+  "The minimum required execution time for a block of repeated trials of a quil program.
+The number of trials will be increased until this target time is exceeded.")
+
+(defun time-quil-prog-raw (q &key (num-trials 1) (num-threads 4))
+  "Return the execution time in seconds of the quil program in qvm Q."
+  (qvm::prepare-for-parallelization num-threads)
+  (simple-time num-trials (run-and-reset-qvm q)))
+
+(defun run-pre-timing (q num-threads init-num-trials)
+  "Perform coarse timing of program in qvm Q."
+  (loop :for num-pre-trials := init-num-trials :then (* num-pre-trials 2)
+        :for trial-time := (time-quil-prog-raw q :num-threads num-threads :num-trials num-pre-trials)
+        :while (< trial-time *minimum-target-time*) ; avoid divide by zero and times very close to zero
+        :finally (return (make-pre-timing :init-num-trials init-num-trials :num-trials num-pre-trials :trial-time trial-time))))
+
+(defun num-trials-in-target-time (pt target-time)
+  "Use coarse timing data in PT to determine the number of trials required to consume TARGET-TIME seconds.
+The latter time is used for more precise timing."
+  (1+ (floor target-time (/ (pre-timing-trial-time pt) (pre-timing-num-trials pt)))))
+
+(defstruct timing
+  "Data characterizing precise (relative to coarse timing) timing of a quil program.
+All slots are meant to be read only, and have no default values."
+  pre-timing
+  num-trials
+  num-threads
+  target-time
+  run-time)
+
+(defun print-timing-info (tm)
+  "Print 1) optimal number of threads; 2) rescaled exectution times for each number of threads
+using the timing data in TM, an instance of TIMING."
+  (let ((*print-pprint-dispatch* (copy-pprint-dispatch)))
+    (set-pprint-dispatch 'float (lambda (s f) (format s "~,2f" f)))
+    (format t "~d: ~a~%" (optimal-num-threads-from-scan tm) (scaled-times-per-trial tm))))
+
+(defun time-per-trial (tm)
+  "Compute number of seconds per thread, per trial from timing data in instance of TIMING struct TM."
+  (/ (timing-run-time tm) (* (timing-num-threads tm) (timing-num-trials tm))))
```

Comment:
I am little shy to ask this next question because I feel I must be missing something basic and will embarrass myself, but they say the one who never asks remains ignorant forever, so here goes...

Why divide by num-threads here?

Suppose I have a program that for fixed nqbits takes exactly `*default-target-time*` = 1s to run 1 trial with 1 thread. In other words

```
time-per-trial = 1 second / (1 thread x 1 trial)
               = 1 sptt (where sptt = seconds/trial-thread)
```

Now suppose that same program still takes 1s to run 1 trial when I increase num-threads to 2, then `time-per-trial = 0.5 sptt`. Since `optimal-num-threads-from-scan` is looking for the number of threads that minimizes the `time-per-trial`, won't it consider 2 to be the optimal number of threads here? Wouldn't 1 be the better choice, since you used fewer resources?

Reply:
Good question. This crossed my mind a few times. I convinced myself this is correct. I'll have another conversation with myself about it...

Reply:
I remember now why I did this. The time measured is actually the total CPU time, not the real time. I found that dividing by the number of threads is an excellent proxy for the wall clock time. Other loads on the machine will make the wall clock time vary, the total CPU time not so much. So it seems robust. I just checked this again, while starting and stopping `yes &> /dev/null`'s.

The documentation should be more clear about this.

Reply:
Of course! I wasn't paying enough attention to your timing macro to notice the difference between `get-internal-run-time` and `get-internal-real-time`! Apologies!  @stylewarning and I just recently had a conversation about these functions, so I have no excuse.

P.S. Did you know that  `get-internal-{run,real}-time` are two of only fourteen external symbols in the `common-lisp` package that are bound to a function whose lambda-list is empty? (Feel free to use this as an icebreaker at your next common-lisp-themed party.)

@stylewarning What good is all this useless trivia if I can't recognize the difference between `get-internal-real-time` and `get-internal-run-time` in actual code?

Reply:
Well, I actually forgot what and why I did it. So, its good to check it and document it.